File: Api/src/main/java/net/winternetwork/bedwars/api/game/stage/Stage.java
package net.winternetwork.bedwars.api.game.stage;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import net.winternetwork.bedwars.api.game.flag.Flag;
import net.winternetwork.bedwars.api.score.ReplaceableList;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;

@Getter
public abstract class Stage {

    private final String stageName;

    private final int time;

    @Setter(value = AccessLevel.PROTECTED)
    private int timeLeft;

    public Stage(String name, int time) {
        this.stageName = name;
        this.time = time;

        this.timeLeft = time;
    }

    public abstract List<Flag> getFlags();

    public abstract ReplaceableList getScoreboard();

    public abstract void onStageJoin();

    public abstract void onSecondPassed();

    public abstract void onTimeLeft();

    public abstract void onStageExit();

    public final void operateTime() {
        this.timeLeft = timeLeft - 1;
    }

    protected final void broadcast(String... messages) {
        for (String message : messages) {
            Bukkit.broadcastMessage(message);
        }
    }

    protected final int getOnlinePlayers() {
        return Bukkit.getOnlinePlayers().size();
    }

    protected final List<Player> getAllPlayers() {
        return new ArrayList<>(Bukkit.getOnlinePlayers());
    }

}

File: Api/src/main/kotlin/net/winternetwork/bedwars/api/game/stage/Stage.kt
package net.winternetwork.bedwars.api.game.stage

import net.winternetwork.bedwars.api.game.flag.Flag
import net.winternetwork.bedwars.api.score.ReplaceableList
import org.bukkit.Bukkit
import org.bukkit.entity.Player

abstract class Stage(
        val name: String,
        val time: Int
) {

    var timeLeft = time
        protected set

    abstract val flags: List<Flag<*>>
    abstract val scoreboard: ReplaceableList

    open fun onStageJoin() {}
    open fun onSecondPassed() {}
    open fun onTimeLeft() {}
    open fun onStageExit() {}

    fun operateTime() {
        timeLeft -= 1
    }

    fun resetTime() {
        timeLeft = time
    }

    protected fun broadcast(vararg messages: String) {
        messages.forEach {
            Bukkit.broadcastMessage(it)
        }
    }

    val onlinePlayers: Int
        get() = Bukkit.getOnlinePlayers().size

    val allPlayers: List<Player>
        get() = ArrayList(Bukkit.getOnlinePlayers())
}